### ISSUE TITLE
DISCUSSION: What should a breaker do with results received during open state, from 'overhanging' calls?

### ISSUE THREAD

**Issue by reisenberger** [App-vNext/Polly#218]
TL;DR: We need to fix a bug where these may currently cause multiple invocations of the `onBreak()` delegate. 

### Scenario

@kharos rightly points out in #216, that a call placed through a circuit-breaker may return to a breaker in a different state from when the call commenced.

While a circuit is open, further call results (from calls initiated before it broke) may still arrive - either successes or failures.

Should the breaker take account of these in metrics?

If a fail arrives on an open circuit, should a Polly circuit pass on that received failure, or throw for broken circuit?

Are there any other aspects to consider? 

### Current behaviour

Bug: If further call failure results arrive on a breaker in open state, the `onBreak()` delegate is triggered a further time for each such call failure.

Metrics governing circuit transitions: the effect of further successes/failures received on an already-broken circuit have no effect on subsequent state transitions.  Transition into `HalfOpen` state is governed by time only.  Transition out of `HalfOpen` state is currently based on a single execution, not metrics.  The circuit resetting (transitioning to `Closed` from `HalfOpen`) wipes the statistics slate clean.   

Reporting metrics (forthcoming): Recommend: record these successes/failures, for completeness. 

Exception behaviour: If a fail arrives on an open circuit, a Polly circuit passes that received failure back to the caller (does not substitute `BrokenCircuitException`).  As in issue #188, this ensures Polly does not mask an outcome actually returned by a call.  Compare the success case: for a success result arriving on an open circuit, it would be similarly (more indeed) misleading to replace a success result with a `BrokenCircuitException`.  This behaviour also preserves the semantics (which some users may be relying on) that a `BrokenCircuitException` indicates the call was never _placed_.   

### Comparison

Hystrix CircuitBreaker: Exception behaviour: as described for Polly above 

Hystrix CircuitBreaker: Metrics governing circuit transitions: as described for Polly above.

Hystrix CircuitBreaker: Reporting metrics: as described for Polly above.

### Recommendation

Fix the bug.  

Preserve exception behaviour.

Change nothing about metrics.  To not 'wipe the metrics slate clean' on re-closing a circuit-breaker, would make systems hard to reason about (for example, overhanging metrics might make it immediately transition to broken again straight after closing).

### Views?

**Comment by Kharos** [comment]
> Bug: If further call failure results arrive on a breaker in open state, the onBreak() delegate is triggered a further time for each such call failure.

Ah, that explains the behavior I could see when Polly allowed two half-open requests (2x onBreak triggered when they both returned).

My expectation is that the event only happens once as I want to use onBreak and the other delegates for logging, allowing us to analyze incidents after-the-fact. I prefer if Polly passes results that it receives in open to the caller normally, as these failures usually have more information for us than the 'broken circuit' exception. So for both of these my expectation seems to match your recommendation.

I don't yet know enough about how Polly's metrics work to comment on that aspect.

**Comment by reisenberger** [comment]
Closing - implemented as recommendation which was set out, in #223
